First version of caller for libs

sebasguts opened the pull request:
This adds a first version for calling all library functions.

## Example for the current looks

```
julia> using Singular

julia> R,(x,) = Singular.PolynomialRing(Singular.QQ, ["x"] )
(Singular Polynomial Ring (QQ),(x),(dp(1),C), spoly{n_Q}[x])

julia> p = 15*x^3+x
15*x^3+x

julia> Singular.LibPoly.normalize(p)
x^3+1/15*x
```

## What is currently in this PR (to be updated)

* A C++-level caller (file `deps/src/caller.cpp`) for library functions, that automatically translates the Julia input into valid Singular input. Currently supported types (all low-level libSingular CxxWrap objects)
  - [x] bigint
  - [x] bigintmat
  - [x] ideal
  - [x] int
  - [x] intmat
  - [x] intvec
  - [x] list
  - [x] map
  - [x] matrix
  - [x] module
  - [x] number
  - [x] poly
  - [x] resolution
  - [x] ring
  - [x] string
  - [x] vector

* A julia level `low_level_caller` (file `src/libsingular/caller.jl`), that calls library functions and translates the output. All high-level Singular.jl types are supported as input (as long as their low-level equivalent is mentioned above). Current supported output is
  - [x] bigint
  - [x] bigintmat
  - [x] ideal
  - [x] int
  - [x] intmat
  - [x] intvec
  - [x] list
  - [x] map
  - [x] matrix
  - [x] module
  - [x] number
  - [x] poly
  - [x] resolution
  - [x] ring
  - [x] string
  - [x] vector
  
* A parser that parses all Singular `.lib` files and creates a julia dictionary. To be hooked into the build script (files `deps/libparser.jl` and `etc/lib_parser.sh`).

* A meta programming part (file `src/libsingular/Meta.jl`) that creates a module for each library and adds all global functions from that library. Note here that all this module generation happens at precompilation time. So no time at the normal startup is wasted.

## What needs to be done

- [x] Support more Singular.jl output types, ideally by making Rings callable on more CxxWrap pointer objects, like for polys:
```
julia> R,(x,) = Singular.PolynomialRing(Singular.QQ, ["x"] )
(Singular Polynomial Ring (QQ),(x),(dp(1),C), spoly{n_Q}[x])

julia> R(x.ptr)
x
```

- [x] Remove some inefficiencies in the code generated by the metapackage (currently, substitution of symbols into a string does not work as expected).
- [ ] Provide some test input
- [x] Add return dictionary for rings
- [x] Add input and output hooks that get executed on the input before calling the function and on the output after calling the functions (for example, to get the right member out of a ring dictionary)
- [x] find a suitable way to rewrap ring pointers
- [ ] Remove the allcaps from library names, probably by prefixing the module name or putting the library modules not directly into singular, but in a separate library module
- [ ] Add a high level wrapper for singulars `bigintmat`, or convert them directly to something in Julia

fingolfin commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,340 @@
+        args[i] = reinterpret_cast<void*>(copy);
+        return true;
+    }
+    if (jl_isa(obj,jl_singular_resolution_type)) {
```
```suggestion
    if (jl_isa(obj, jl_singular_resolution_type)) {
```

sebasguts replied:
I already removed this whole function . . .

fingolfin commented on etc/parse_libs.sh:
```diff
@@ -0,0 +1,5 @@
+
+## create json for all files
+
+/home/sebastian/Software/Singular-git/Singular/libparse -f $1 | awk -F " " '{print "[\"" $1 "\",\"" $3 "\"]\,"}' | tail -n+4
```
That seems rather specific to your system ;-)

sebasguts replied:
Indeed. Work in progress atm ;)

fingolfin commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,42 @@
+    )
+end
+
+parsepath = abspath(joinpath(@__DIR__,"..","etc","parse_libs.sh"))
```
`parse_libs.sh` seems to be only there to invoke `<PATH>/Singular/libparse`, and then pass its output through awk and tail. That seems needlessly complicated. Why not call `libparse` directly, and do the parsing in Julia? 

On the other hand, `parselibs.jl` doesn't seem to be used (yet?)...

sebasguts replied:
It is WIP. I will set the paths correctly. Unfortunately, I had to rip apart my `build.jl`, so I did not commit changes in it.

sebasguts replied:
Well, it was somehow more convenient for me to make the initial parsing with `awk` and not with Julia, as I have no idea how to handle `CSV`-style stuff with Julia. I could probably do without the shell script, though.

fingolfin commented on src/LibSingular.jl:
```diff
@@ -4,8 +4,14 @@ import Libdl
 function __init__()
    @initcxx
+   global mapping_types, casting_functions
+   mapping_types = get_type_mapper()
```
Really? :-) I'd at least expect consistency between the variable and the getter, e.g. `type_mapper = get_type_mapper()` or `mapping_types = get_mapping_types`.

fingolfin commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,340 @@
+        std::pair<int, std::string>(BIGINTMAT_CMD, "BIGINTMAT_CMD"),
+        std::pair<int, std::string>(MAP_CMD, "MAP_CMD"),
+        std::pair<int, std::string>(RESOLUTION_CMD, "RESOLUTION_CMD"),
+        std::pair<int, std::string>(INTVEC_CMD, "INTVEC_CMD")};
```
Wow, that's horribly inefficient. Though of course you only call `get_type_mapper` once, so it doesn't matter... but have you considered just doing this instead?
```C
struct { int cmd; const char *name; } types = {
    { BIGINT_CMD, "BIGINT_CMD" },
    ...
    { INTVEC_CMD, "INTVEC_CMD" },
};
```
And then replace `types.size()` by `sizeof(types)/siezof(types[0])`, `types[i].first` by `types[i].type` and `types[i].second.c_str()` by `types[i].name`.

fingolfin replied:
I see that various types are missing, but of course that's fine since this is just a prototype anyway :-).

sebasguts replied:
It is only for the initial setup. I kinda want to replace it by something generated in the future.

sebasguts replied:
Added that (minus smaller typos in your suggestion), but disabled clang-format for it, as it would put several entries in one line.

fingolfin commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,340 @@
+    jl_singular_bigint_type = jl_get_global(module,jl_symbol("__mpz_struct"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("bigintmat"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("sip_smap"));
+    jl_singular_resolution_type = jl_get_global(module,jl_symbol("resolvente"));
```
Totally unrelated to this PR, but: huh, what is a "resolvente"? Or is that "denglish" / "germanglish" for "resolvent"? But even then, I never heard the term "resolvent" used as an alias for "resolution" in this context (googling reveals such a use in logic, but not in commutative algebra). But of course I am hardly an expert, so this is not exactly a surprise; I'd still be curious to learn about a reference or two where this word is used like this.


Uhh, I see that this is defined in Singular, and also has been exposed under the same name in `Singular.jl`.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,340 @@
+    return result;
+}
+
+intvec* jl_array_to_intmat(jl_value_t* array_val){
```
Yeah, there is one. I just did not apply it for the latest commit, that is why something is formatted and something is just arbitrary. I will format the whole document for my next commit.

sebasguts commented:
Even though this might have some inconsistencies, I think it is ready for review now.

sebasguts commented:
@hannes14 Can you have a look at the tests failing for the resolutions?

I have fixed most, the other ones are tests for length, which has changed with our new setup. Do we return the length incorrectly?

fingolfin commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,391 @@
+    Singular.method("safe_singular_string", &safe_singular_string);
+
+    Singular.method("get_length_of_resolution", &get_length_of_resolution);
+    Singular.method("get_resolution_data", &get_resolution_data);
```
The above two functions are never called. Intentional? Perhaps at least a test case should call them?

sebasguts replied:
No, it is also a leftover. Will remove it.

fingolfin commented on deps/src/coeffs.cpp:
```diff
@@ -44,9 +44,8 @@ void singular_define_coeffs(jlcxx::Module & Singular)
-    Singular.method("n_Delete_Q", [](void * n, coeffs cf) {
-        number   tt = reinterpret_cast<number>(n);
-        number * t = &tt;
+    Singular.method("n_Delete_Q", [](snumber * n, coeffs cf) {;
```
```suggestion
    Singular.method("n_Delete_Q", [](snumber * n, coeffs cf) {
```
BTW, all places using `snumber *` could also just use `number`

sebasguts replied:
I would say it works, otherwise the tests would segfault.

fingolfin commented on deps/src/coeffs.cpp:
```diff
@@ -44,9 +44,8 @@ void singular_define_coeffs(jlcxx::Module & Singular)
+    Singular.method("n_Delete_Q", [](snumber * n, coeffs cf) {;
+        number * t = &n;
         if (n != NULL) {
             n_Delete(t, cf);
```
Why introduce that helper variable at all? You can just do this:
```suggestion
            n_Delete(&n, cf);
```

sebasguts replied:
Classic no-brainer actually. it is a leftover when there was still a cast from `void*`

fingolfin commented on deps/src/singular.cpp:
```diff
@@ -52,65 +54,73 @@ JLCXX_MODULE define_julia_module(jlcxx::Module & Singular)
-        auto ra = reinterpret_cast<resolvente>(ra_void);
-        return (ideal)ra[k];
+    Singular.method("getindex_internal", [](syStrategy ra, int64_t k, bool minimal) {
+        if(minimal){
```
```suggestion
        if (minimal) {
```
BTW, there is also a `.clang-format` style file, so you could probably just use `git clang-format` on this?

sebasguts replied:
Will do so.

fingolfin commented on test/caller-test.jl:
```diff
@@ -0,0 +1,46 @@
+    @test Singular.SETS.isEqualInt(R, "aa", "aa")                     == 1
+    @test Singular.SETS.isEqualInt(R, Singular.QQ(1), Singular.QQ(1)) == 1
+    @test Singular.SETS.isEqualInt(R, [1,2,3], [1,2,3])               == 1
+    @test Singular.SETS.isEqualInt(R, [1 2; 3 4], [1 2; 3 4])         == 1
```
Perhaps rewriting these tests with the order of the `==` arguments switched would be easier to maintain? I.e.
```suggestion
    @test 1 == Singular.SETS.isEqualInt(R, [1 2; 3 4], [1 2; 3 4])
```
But of course it's fine as it is, too

wbhart commented on test/caller-test.jl:
```diff
@@ -0,0 +1,46 @@
+
+    vec = FreeModule(R,2)([x,y])
+    mod = Singular.Module(R, vec)
+    i1 = Singular.POLY.mod2id(R,mod,[1,2])
```
I'm not particularly fond of all CAPS names. It is not done in Julia in general.

sebasguts replied:
The main problem is that I could not add for example a `Poly` submodule to the `Singular` module, because Module names need to be unique and cannot be types or whatever. So whichever other (reasonable) spelling I tried (all small, first letter capitalized) it always failed. This is why I did all-caps. The only solution would be to put the functions in a different module, say `Libs`. But then the names get longer. If you have a different suggestion, it is highly welcome.

sebasguts replied:
Ping on the naming stuff here. Did anybody come up with a better naming convention?

wbhart commented on src/LibSingular.jl:
```diff
@@ -1,11 +1,16 @@
 import Libdl
+
+# We need the top-level eval for now
+import ..Singular
```
Can you explain where you are using this. Normally I would say that this indicates that you have not got a clean separation between libsingular and Singular namespaces. But as you added a comment, I assume this was actually necessary.

sebasguts replied:
I think it is also a leftover and can be removed. The caller files are mostly removed from the `LibSingular` module anyway.

wbhart commented:
By the way, when you are "using Singular", all the symbols exported in Singular are available at the top level. You don't need to write Singular.PolynomialRing or Singular.QQ, etc. In fact, it is bad practice to do this, because it hides cases where things should have been exported for the user, but have accidentally been omitted from the export list.

wbhart commented on test/caller-test.jl:
```diff
@@ -0,0 +1,46 @@
+    # i1 = Ideal(R, x*z, y*z, x^3-y^3)
+    # @test Singular.STANDARD.res(R,i1,0) isa Singular.sresolution
+    # i1 = Ideal(R, x*z, y*z, x^3-y^3)
+    # @test Singular.PRIMDEC.primdecGTZ(R,i1) isa Array
```
Is there any reason these tests are commented out? Wolfram particularly wants this to work, so it would be great if this could be enabled, especially as it will give a convenient example for users, given that there are no docs for the Singular libraries in the Singular.jl docs.

sebasguts replied:
They were failing last week because Singular was buggy. I will reenable them, because they should pass now.

sebasguts commented on deps/build.jl:
```diff
@@ -93,9 +93,9 @@ end
+      run(`$srcs/configure --with-libparse --prefix=$vdir --disable-static --enable-p-procs-static --disable-p-procs-dynamic --disable-gfanlib --enable-shared --with-gmp=$nemovdir --with-flint=$nemovdir --with-ntl=$vdir --without-python --with-readline=no`)
    else
-      run(`$srcs/configure --prefix=$vdir --disable-static --enable-p-procs-static --disable-p-procs-dynamic --disable-gfanlib --enable-shared --with-gmp=$nemovdir --with-flint=$nemovdir --with-ntl=$vdir --without-python --with-readline=no --with-debug --enable-debug --disable-optimizationflags`)
+      run(`$srcs/configure --with-libparse --prefix=$vdir --disable-static --enable-p-procs-static --disable-p-procs-dynamic --disable-gfanlib --enable-shared --with-gmp=$nemovdir --with-flint=$nemovdir --with-ntl=$vdir --without-python --with-readline=no --with-debug --enable-debug --disable-optimizationflags`)
```
Yep. But #109 currently fails (needs some investigation)

fingolfin replied:
Now #109 is merged, so this needs a rebase

sebasguts commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,43 @@
+    (
+      stdout = String(read(out)), 
+      stderr = String(read(err)),  
+      code = process.exitcode
```
I could try to check if there is something on `stderr`, but I am not sure I can really do sensible error handling, or just return stderr.

sebasguts commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,43 @@
+if haskey(ENV,"SINGULAR_LIBRARY_DIR")
+    library_dir = ENV["SINGULAR_LIBRARY_DIR"]
+else
+    library_dir = abspath(joinpath(@__DIR__,"..","local","share","singular","LIB"))
```
Good plan.

sebasguts commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,43 @@
+
+filenames = filter(x -> endswith(x,".lib"),readdir(library_dir))
+
+output_filename = abspath(joinpath(@__DIR__,"..","src","libraryfuncdictionary.jl"))
```
I really prefer to make everything absolute. It is not needed per se, but it also does not hurt.

sebasguts commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,43 @@
+
+open(output_filename,"w") do outputfile
+    println(outputfile,"""
+libraryfunctiondictionary = Dict(""")
```
Will do.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,340 @@
+    int *        result_content = result->ivGetVec();
+    for (int i = 0; i < size; i++) {
+        result_content[i] =
+            static_cast<int>(jl_unbox_int64(jl_arrayref(array, i)));
```
Well, making it only work with `Int32` is IMHO out of question, as this will be super inconvenient. I could provide both versions, as @fingolfin suggests, and raise an error if there is an overflow. I can either check on the Julia or the C++-side, it does not matter.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,391 @@
+    return *reinterpret_cast<void **>(obj);
+}
+
+// Safe way
```
Depends on the setup. For my editor, this comment style is more natural.

wbhart replied:
Ok.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,391 @@
+    return reinterpret_cast<void *>(result);
+}
+
+static bool check_vector(poly p, ring r)
```
Na, @fingolfin is talking about `check_vector`. It is indeed unused, just a leftover from previous versions. I will remove it.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,391 @@
+static void * get_ring_ref(ring r)
+{
+    r->ref++;
+    return reinterpret_cast<void *>(r);
```
The function call decreases it automatically.

fingolfin replied:
Sorry, I don't understand: which function call does decrease it?

sebasguts replied:
I will add a comment. The function call to a library function destroys its arguments. We increase the reference, as the call will decrease it.

sebasguts commented on deps/src/caller.cpp:
```diff
@@ -0,0 +1,391 @@
+    return reinterpret_cast<void *>(r);
+}
+
+static jl_value_t * get_poly_ptr(poly p, ring r)
```
I will change the name to something like `copy_polyptr_to_void`

sebasguts commented on src/Meta.jl:
```diff
@@ -0,0 +1,34 @@
+    )
+)
+
+for (name,funcs) in libraryfunctiondictionary
```
Will do.

fingolfin replied:
It doesn't do much at all yet though, does it? Given that the  `input/output_manipulator_fucncs` dicts are essentially empty. Perhaps more code is meant to be added here eventually?

sebasguts replied:
Well, it still creates the library modules and functions, so it creates some functions, but they just call the library functions, nothing more. @jankoboehm wanted to fill those dictionaries over time.

sebasguts commented on src/module/ModuleTypes.jl:
```diff
@@ -32,6 +32,10 @@ mutable struct svector{T <: Nemo.RingElem} <: Nemo.ModuleElem{T}
    end
 end
 
+function (R::PolyRing{T})(m::libSingular.poly,::Val{:vector}) where T
```
It is internal. Do we want to document it, really? I think it would be better if it is not used outside the caller.

wbhart replied:
The docs are developer documentation, so at the very least, put a comment saying what the functions do, even if you don't put a docstring for the docs.

sebasguts commented on src/module/ModuleTypes.jl:
```diff
@@ -88,6 +92,10 @@ mutable struct smodule{T <: Nemo.RingElem} <: Module{T}
    end
 end
 
+function (R::PolyRing{T})(m::libSingular.ideal,::Val{:module}) where T
```
Same as for vector

sebasguts commented:
> What I am not convinced of is the need to call such functions via Singular.SOMENAMEINCAPSLIKEGAP.function. This doesn't just create a bunch of top-of-module-level Julia functions that apply to the existing high level objects that we had before in Singular.jl. For example, in Singular, when I import a library, those functions get injected into scope and I can just call them on interpreter objects directly. (Or at least I think it does; just checking this with Hans now.)

I do not really understand that comment can you elaborate on what is bothering you about it in particular?

wbhart commented:
> > What I am not convinced of is the need to call such functions via Singular.SOMENAMEINCAPSLIKEGAP.function. This doesn't just create a bunch of top-of-module-level Julia functions that apply to the existing high level objects that we had before in Singular.jl. For example, in Singular, when I import a library, those functions get injected into scope and I can just call them on interpreter objects directly. (Or at least I think it does; just checking this with Hans now.)
> 
> I do not really understand that comment can you elaborate on what is bothering you about it in particular?

It's not convenient for the user. Who wants to type all of that? Remember Singular.jl is designed to be used standalone, not just as a backend for something else.

The capital letters really have to be changed. Consider LibPoly instead of POLY, assuming that is what the name of the Singular library is. It's fine to be able to access the functions via this, but the user will not want to do this.

However, let's see if @hannes14 has an opinion on this. Does the Singular interpreter force you to write Poly.blah to call the function blah in the Poly library, or does it just inject the function names so you can just type blah?

Maybe some of the function names already exist because we already implemented them directly via the kernel? I certainly prefer that rather than have them handled via the interpreter. LibPoly sounds like default Singular functionality rather than functionality from a library. I guess I don't mind default Singular functionality being accessible via Interpreter.blah, for the cases where we haven't wrapped the corresponding kernel function. But surely true library functionality should be imported for the user.

sebasguts commented on deps/src/CMakeLists.txt:
```diff
@@ -10,10 +10,10 @@ include_directories(${nemo_includes})
 include_directories(${singular_includes}/singular)
 
-SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14" )
+SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -g" )
```
Exactly. I really want to run this code sensible in the debugger, set breakpoints, etc.

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,191 @@
+
+function prepare_argument(x::sideal)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
```
Why is `rng_ptr` needed here and elsewhere? It's not shorter than `rng.ptr`...

fingolfin commented on src/resolution/resolution.jl:
```diff
@@ -21,13 +21,13 @@ elem_type(::ResolutionSet{T}) where T <: AbstractAlgebra.RingElem = sresolution{
 
 function checkbounds(r::sresolution, i::Int)
-   (i < 1 || i > r.len) && throw(BoundsError(I, i))
+   (i < 1 || i > libSingular.get_sySize(r.ptr)) && throw(BoundsError(r, i))
```
```suggestion
   (i < 1 || i > length(r)) && throw(BoundsError(r, i))
```

sebasguts commented:
> Consider LibPoly instead of POLY, assuming that is what the name of the Singular library is. It's fine to be able to access the functions via this, but the user will not want to do this.

I changed it to `LibPoly`

> Maybe some of the function names already exist because we already implemented them directly via the kernel?

Absolutely not the problem. `Ideal` is defined as the constructor for an ideal in Singular.jl, but it is also a Library. Some functions have the same name as their library name. This has nothing to do with distinction between kernel and library, but that library names are orthogonal to anything. And Singular users know in which library their stuff lives. Exporting single library functions into the Singular module is possible, but this needs to be done by hand. Currently, putting each library into a different module is the right choice.

fingolfin commented on src/resolution/resolution.jl:
```diff
@@ -98,13 +103,16 @@ end
-   if r.len > 0
-      ptr = libSingular.getindex(r.ptr, Cint(0))
-      print(io, "R^", libSingular.rank(ptr))
+   len = libSingular.get_sySize(r.ptr)
```
```suggestion
   len = length(r)
```

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,191 @@
+# end
+
+casting_functions_pre = Dict(
+    :NUMBER_CMD     => (libSingular.NUMBER_CMD_CASTER,     true  , ()),
```
A comment explaining the meaning of the entries of those tuples would be nice

fingolfin commented on deps/src/coeffs.cpp:
```diff
@@ -44,11 +44,10 @@ void singular_define_coeffs(jlcxx::Module & Singular)
-        number   tt = reinterpret_cast<number>(n);
-        number * t = &tt;
+    Singular.method("n_Delete_Q", [](snumber * n, coeffs cf) {
+        ;
```
```suggestion
```

fingolfin commented on etc/parse_libs.sh:
```diff
@@ -0,0 +1,5 @@
+
+## create json for all files
+
+$1 -f $2 | awk -F " " '{print "[\"" $1 "\",\"" $3 "\"]\,"}' | tail -n+4
```
Please add trailing new line. 

Also, I don't see how this is simpler than doing it in Julia directly? All `awk` does here is to split each line at spaces, which is trivial to do in julia; and then it assembles it together into some JSON.

I don't know right now what the input data to be parsed here looks like, but  if you give me a sample input and output, I'd be happy to provide a Julia version of it.

fingolfin commented on src/Meta.jl:
```diff
@@ -0,0 +1,34 @@
+        if i[1] == "g"
+            func_name = i[2]
+            symb = Symbol(func_name)
+            input_manipulator = haskey(input_manipulator_funcs,name) && haskey(input_manipulator_funcs[name],symb) ? input_manipulator_funcs[name][symb] : identity
```
These lines are way to full. I'd suggest to choose shorter variable names. Also could use Dict::get, like so (untested), or perhaps replace the `?:` operator by a plain `if`?
```suggestion
            input_manipulator = haskey(input_manipulator_funcs,name) ? get(input_manipulator_funcs[name], symb, identity) : identity
```

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+
+function convert_return_value(single_value, rng = nothing)
+    if single_value[1]
+        error("recieved list instead of single value")
```
```suggestion
        error("received list instead of single value")
```

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+    return return_dict
+end
+
+function convert_return_value(single_value, rng = nothing)
```
Some comments here and elsewhere are badly missing. For starters, what is `single_value` expected to contain?

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+# end
+
+casting_functions_pre = Dict(:NUMBER_CMD     => (libSingular.NUMBER_CMD_CASTER,     true, ()),
+    :RING_CMD       => (libSingular.RING_CMD_CASTER,       false, ()),
```
A comment explaining the meaning of the entries of those tuples would be nice

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+#     x_new = LIST_CMD_TRAVERSAL(x)
+# end
+
+casting_functions_pre = Dict(:NUMBER_CMD     => (libSingular.NUMBER_CMD_CASTER,     true, ()),
```
Looking at this, I guess I don't like the VS code auto-formatter that much :-/

sebasguts replied:
Yeah, it produces terribly ugly code. On the other hand, it is a standard. I am not sure which I like more, tbh.

wbhart replied:
You must have an option set incorrectly. It can't be a standard to use single line comments everywhere. I use VSCode and I use multiline comments.

sebasguts replied:
That has nothing to do with the comments, but with the dict input formatting. I put the single line comments, because that is my setting, and it is not incorrect at all.

However, the code formatting does not allow for settings. So here we are.

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+    arguments = Any[ i for (i, j) in arguments ]
+    return_values = nothing
+    if rng == nothing
+        return_value = libSingular.call_singular_library_procedure(name, arguments)
```
How about this instead (assuming it works as intended, which it may not, I don't know anything about cxxwrapper):
```suggestion
        return_value = libSingular.call_singular_library_procedure(name, C_NULL, arguments)
```

Then you can get rid of `call_singular_library_procedure_wo_ring` (I find the result easier to grok, but of course that's highly subjective, so i won't complain if you leave it as is ;-)

In fact, I'd then just do this:
```julia
   rng = (rng == nothing) ? C_NULL : rng.ptr
   result = libSingular.call_singular_library_procedure(name, rng, arguments)
   return convert_return_list(result, rng)
end
```

sebasguts replied:
Unfortunately, this won't work as easy as you think here.

Because `call_singular_library_procedure` is a CxxWrapped function, the argument needs to be of the right type. There is no way we could either give a `ring` or a `C_NULL`, which is a `Ptr{Nothing}`.

We would have the following possibilities:
1. Pass the raw ring pointer to `call_singular_library_procedure` instead of the ring object. This might cause trouble if the CxxWrap's object structure ever changes (although quite some code relies on this anyway).
2. We overload this function to either take `void*` or `ring`.

I implemented the second for now.

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+end
+
+function prepare_argument(x::Array{Int64,1})
+    return Any[ mapping_types_reversed[:INTVEC_CMD], libSingular.jl_array_to_intvec(x) ], nothing
```
I find this whole business of generating an array in the C code in function `get_type_mapper `, exporting this to Julia, then turning it into a "reverse" mapping dictionary `mapping_types_reversed` there, then using that to convert a Julia symbol to an integer on the Julia level, then passing this back to the kernel, overall somewhat hard to track.

Can't we radically simplify this, by creating that "reverse mapping" dictionary in the kernel (possibly even as a plain C++ map!), and never exporting it from there? Then the above line would become just this:
```suggestion
    return Any[ :INTVEC_CMD, libSingular.jl_array_to_intvec(x) ], nothing
```
And in `caller.cpp` code, you only need to modify `translate_singular_type`, to expect a julia symbol as first argument, instead of integer. It then only needs to modify that integer.

I believe with this relatively straightforward change, things get quite a bit simpler.

fingolfin replied:
(I missed that also `create_casting_functions` and friends would need to be updated).

Anyway, thinking some more about this, it would probably be even simpler if indeed some of this stuff was replaced by a program that generates the required data (just as we do in https://github.com/gap-packages/SingularInterface) -- then that C/C++ program could in particular creates a `.jl` file which simply defines Julia constants with the values of `INTVEC_CMD` etc.. Then, we can use these everywhere instead of symbols, and completely dispense with the conversions between symbols and ints. Also, Julia could then optimize the code some more, I guess.

fingolfin replied:
Looking at the CxxWrap.jl "documentation" (i.e., its README -- is there anything else?), I guess one could also export the `yytokentype` enum to Julia. But from the explanations there, it is unclear to me whether one can convert the enum values into an `Cint`, as we would want.

wbhart replied:
I think we should open a ticket for this and deal with it later. I doubt Sebastian has time to work on such a thing now.

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+function low_level_caller_rng(lib::String, name::String, ring, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    arguments = Any[ i for (i, j) in arguments ]
```
So you throw away the rings here. Is this safe? What will happen if I pass in an argument that lives in a different ring? At least in the past, this could crash Singular, IIRC (but perhaps I am wrong? @hannes14 will know better :-)

sebasguts replied:
Ideally, we would compare rings, yes.

fingolfin commented on src/caller.jl:
```diff
@@ -0,0 +1,189 @@
+    rng = nothing
+    for (i, j) in arguments
+        if j != nothing
+            rng = j
```
So, you take the last ring of all the rings. Why not the first (and then insert a `break` here)? Or, even better: Why not check that all subsequent rings are compatible with the first -- as I also suggested for other variant of this function. With pseudo code like this:
```julia
arguments = []
for i in args
  i, j = prepare_argument(i)
  push!(arguments, i)
  if j != nothing
    if rng == nothing
      rng = j
    else
      # TODO: check "compatibility, e.g. rng.ptr == j.ptr
      # (but perhaps that is too restrictive?)
    end
  end
```
With that, the two function variants can be merged, with this ring-free variant becoming just `return low_level_caller(lib, name, nothing, args...)`

fingolfin commented on src/ideal/ideal.jl:
```diff
@@ -495,16 +495,8 @@ function sres(I::sideal{T}, max_length::Int) where T <: Nemo.RingElem
-   end
-   return sresolution{T}(R, length, r, minimal)
+   r, minimal = libSingular.id_sres(I.ptr, Cint(max_length + 1), R.ptr)
+   return sresolution{T}(R, r, minimal)
```
I am wondering: in how far are this change and the many others similar ones tied to our library calling work? Perhaps they could / should be separated? (Just wondering -- of course that is extra work, and so we may not want it; the main reason I bring it up is that this PR is huuuge and it's typically easier to review, test and merge multiple smaller PRs, as you of course know -- it's a balance between your time, and that of reviewers :-)

wbhart replied:
Let's just get the whole thing merged. Don't bother splitting it.

fingolfin commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,46 @@
+        println(outputfile, "],\n")
+    end
+    println(outputfile, ")\n")
+end
```
This code doesn't look so bad to me, and has the advantage that to understand it, one just needs to know Julia, while in the old setup, one had to know `awk` as well :-)

fingolfin commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,46 @@
+        if libs.stderr != ""
+            error("from libparse: $(libs.stderr)")
+        end
+        libs_splitted = split(libs.stdout,"\n")[4:end-1]
```
I think a comment here would be very helpful; something like this:
```julia
# The output of libparse looks as follows: each line has the form ...
#  .... give generic description of the form ...
#  For a concrete example, it might be this:
#  ... concrete example ...
#  From this, we need to extra ...
```

fingolfin commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,46 @@
+        end
+        libs_splitted = split(libs.stdout,"\n")[4:end-1]
+        libs_splitted = [ split(i," ") for i in libs_splitted ]
+        libs_splitted = [ [ j for j in i if j != ""] for i in libs_splitted ]
```
This is redundant if you add `keepempty = false` to your call to `split`

fingolfin commented on deps/parselibs.jl:
```diff
@@ -0,0 +1,46 @@
+            error("from libparse: $(libs.stderr)")
+        end
+        libs_splitted = split(libs.stdout,"\n")[4:end-1]
+        libs_splitted = [ split(i," ") for i in libs_splitted ]
```
The next line becomes redundant if you do this instead:
```suggestion
        libs_splitted = [ split(i," ", keepempty = false) for i in libs_splitted ]
```

Also, it seems you prefer list comprehensions (which do me is very Pythonic), while I prefer classic FP style map/filter/... (probably because it is close to what I'd do in GAP). I have no idea what is considered to be "more typical Julia" or preferred in this project; perhaps it doesn't matter. But just to mention it, I think you could also do this (but this is not a change request):
```suggestion
        libs_splitted = map( i-> split(i," ", keepempty = false), libs_splitted ]
```

fingolfin commented:
It would be a pity if this work bitrotted. I've thus now rebased this, and squashed it into one commit; then added one commit with a few minor changes that address a few of my own remarks (it can be squashed later, I just wanted to make transparent what I did).

I wonder if perhaps this PR could be split into some more digestible parts; e.g. the syzygy related changes seem as if they might be useful independently of the library caller stuff?

wbhart commented:
It seems to come close to passing, except for the off by one error in resolutions and a missing function.

If we can get it passing, we should just merge it. Singular.jl is not so mature as to agonise over every PR.

sebasguts commented:
@fingolfin Your changes are fine. As for the resolutions, we should probably send an email to Hans (I also already tagged him somewhere in some comment) since I have no idea if the changes or the tests are actually wrong. I think it is mostly an off by one error, that is strangely deep down in some Singular data structure. I will send him an email, asking about this.

sebasguts commented:
I added some comments and pushed some tweaks. More to come later.

wbhart commented:
I asked @hannes about the length failures and it seems we should subtract 1 from the length as in the original code. I've made these changes, which should kill 6 of the 8 test failures.

wbhart commented:
I propose we add tickets for the remaining three todos and work on fixing those at some later date. People need access to this as soon as possible, so we should get it working rather than letting it bitrot.

wbhart commented:
I'm going to merge this now, as we are about to have some breaking changes in Singular.jl, and I don't fancy rebasing this patch. We can argue about the definition of a resolution later and change it if necessary.

As far as I see, there are just the following changes that need to be added in future:

* Additional code comments as requested by @fingolfin
* Implement bigintmat
* Provide some test input (though I have no idea what this means)
* Remove ALLCAPS from library names

I will eventually open tickets for these, as @sebasguts has only volunteered to do the first of these.
